lightning/src/ln/script.rs

```diff
@@ -0,0 +1,285 @@
+//! Abstractions for scripts used in the Lightning Network.
+
+use bitcoin::bech32::u5;
+use bitcoin::blockdata::opcodes::all::OP_PUSHBYTES_0 as SEGWIT_V0;
+use bitcoin::blockdata::script::Script;
+use bitcoin::hashes::Hash;
+use bitcoin::hash_types::{PubkeyHash, ScriptHash, WPubkeyHash, WScriptHash};
+use bitcoin::secp256k1::key::PublicKey;
+
+use ln::features::InitFeatures;
+use ln::msgs::DecodeError;
+use util::ser::{Readable, Writeable, Writer};
+
+use std::convert::From;
+use std::convert::TryFrom;
+use std::io::Read;
+
+/// A script pubkey for shutting down a channel as defined by [BOLT #2].
+///
+/// [BOLT #2]: https://github.com/lightningnetwork/lightning-rfc/blob/master/02-peer-protocol.md
+#[derive(Clone)]
+pub struct ShutdownScript(ShutdownScriptImpl);
+
+/// An error occurring when converting from [`Script`] to [`ShutdownScript`].
+#[derive(Debug)]
+pub struct InvalidShutdownScript(Script);
+
+#[derive(Clone)]
+enum ShutdownScriptImpl {
+	/// [`PublicKey`] used to form a P2WPKH script pubkey. Used to support backward-compatible
+	/// serialization.
+	Legacy(PublicKey),
+
+	/// [`Script`] adhering to a script pubkey format specified in BOLT #2.
+	Bolt2(Script),
+}
+
+impl Writeable for ShutdownScript {
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), ::std::io::Error> {
+		self.0.write(w)
+	}
+
+	fn serialized_length(&self) -> usize {
+		self.0.serialized_length()
+	}
+}
+
+impl Readable for ShutdownScript {
+	fn read<R: Read>(r: &mut R) -> Result<Self, DecodeError> {
+		Ok(ShutdownScript(ShutdownScriptImpl::read(r)?))
+	}
+}
+
+impl_writeable_tlv_based_enum!(ShutdownScriptImpl, ;
+	(0, Legacy),
+	(1, Bolt2),
+);
+
+impl ShutdownScript {
+	/// Generates a P2PKH script pubkey from the given [`PubkeyHash`].
+	pub fn new_p2pkh(pubkey_hash: &PubkeyHash) -> Self {
+		Self(ShutdownScriptImpl::Bolt2(Script::new_p2pkh(pubkey_hash)))
+	}
+
+	/// Generates a P2SH script pubkey from the given [`ScriptHash`].
+	pub fn new_p2sh(script_hash: &ScriptHash) -> Self {
+		Self(ShutdownScriptImpl::Bolt2(Script::new_p2sh(script_hash)))
+	}
+
+	/// Generates a P2WPKH script pubkey from the given [`WPubkeyHash`].
+	pub fn new_p2wpkh(pubkey_hash: &WPubkeyHash) -> Self {
```

Comment:
Does it make sense to have a constructor that takes the `PublicKey` so that we can construct backwards-compatible objects? I dunno if we strongly care for now, but I think in the future we should ensure that we always have a compatible API which allows users to write objects which the immediately-previous version can read so that they can upgrade at their own pace (including with users running multiple versions reading the same node data).

Reply:
I implemented `From<PublicKey>` for this case, which is used in `Channel`'s implementation of `ReadableArgs`, so I think this is already working as you described it should.

Reply:
Ah, I don't usually look at the `From`s super closely. Should we have a `new_p2wpkh_explicit` or something? We should at least note in the changelog that you have to use the `From` if you want backwards compat.